staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go

```diff
@@ -101,55 +68,100 @@ func WithAudit(handler http.Handler, requestContextMapper request.RequestContext
 			responsewriters.InternalError(w, req, errors.New("no context found for request"))
 			return
 		}
-		attribs, err := GetAuthorizerAttributes(ctx)
-		if err != nil {
-			responsewriters.InternalError(w, req, err)
-			return
+
+		// prefer the id from the headers. If not available, create a new one.
+		// TODO(security): do we want to forbid the header for non-front-proxy users?
+		var id string
+		ids := req.Header[AuditIDHeader]
+		if len(ids) > 0 {
+			id = ids[0]
+		} else {
+			id = uuid.NewRandom().String()
 		}
 
-		username := "<none>"
-		groups := "<none>"
-		if attribs.GetUser() != nil {
-			username = attribs.GetUser().GetName()
-			if userGroups := attribs.GetUser().GetGroups(); len(userGroups) > 0 {
-				groups = auditStringSlice(userGroups)
+		// set the level
+		level := auditinternal.LevelMetadata
+		if policy != nil && len(policy.Rules) > 0 {
+			// This is just a hack to get through the test without setting a high level by default.
+			// TODO: add the policy evalutation here
+			level = policy.Rules[0].Level
+		}
+		headerLevels := req.Header[AuditLevelHeader]
```

Comment:
I wasn't expecting this.  Something we agreed to in the proposal?  This would allow a user to make his request a lot more expensive to serve (assuming we only allow him to ratchet up) and as implemented would allow him to skip it entirely, right?